### ISSUE TITLE
Build failure: haskell.compiler.ghc{8107,902,92*,94*} on *-darwin

### ISSUE THREAD

**Issue by sternenseemann** [NixOS/nixpkgs#367686]
Almost certainly due to the LLVM 16 -> 19 upgrade in #361878. LLVM 18 seems to have added the following check: https://github.com/llvm/llvm-project/commit/d506aa4edfa66074db3dc1fa84da9d9c80d71500. It's not yet clear to me how GHC >= 9.6 (with hadrian doesn't trip that issue).

```
"inplace/bin/ghc-stage1" -optc--target=arm64-apple-darwin -optc-Wall -optc-Wall -optc-Wextra -optc-Wstrict-prototypes -optc-Wmissing-prototypes -optc-Wmissing-declarations -optc-Winline -optc-Wpointer-arith -optc-Wmissing-noreturn -optc-Wnested-externs -optc-Wredundant-decls -optc-Wno-aggregate-return -optc-Wno-unused-label -optc-Iincludes -optc-Iincludes/dist -optc-Iincludes/dist-derivedconstants/header -optc-Iincludes/dist-ghcconstants/header -optc-Iincludes/dist-install/build -optc-Irts -optc-Irts/dist/build -optc-DCOMPILING_RTS -optc-DFS_NAMESPACE=rts -optc-fno-strict-aliasing -optc-fno-common -optc-Irts/dist/build/./autogen -optc-Wno-unknown-pragmas -optc-O2 -optc-fomit-frame-pointer -optc-g -optc-DRtsWay=\"rts_v\" -optc-w -static  -H32m -O -Wall  -Iincludes -Iincludes/dist -Iincludes/dist-derivedconstants/header -Iincludes/dist-ghcconstants/header -Iincludes/dist-install/build -Irts -Irts/dist/build -DCOMPILING_RTS -DFS_NAMESPACE=rts -this-unit-id rts -dcmm-lint     -package-env - -i -irts -irts/dist/build -Irts/dist/build -irts/dist/build/./autogen -Irts/dist/build/./autogen            -O2 -Wcpp-undef   -Wnoncanonical-monad-instances  -c rts/StgCRun.c -o rts/dist/build/StgCRun.o

/private/tmp/nix-build-ghc-9.2.8.drv-0/ghc42576_0/ghc_1.s:54:1: error:
     error: non-private labels cannot appear between .cfi_startproc / .cfi_endproc pairs
   |
54 | _StgReturn:
   | ^
_StgReturn:
^

/private/tmp/nix-build-ghc-9.2.8.drv-0/ghc42576_0/ghc_1.s:12:2: error:
     error: previous .cfi_startproc was here
            .cfi_startproc
            ^
   |
12 |         .cfi_startproc
   |  ^
`cc' failed in phase `Assembler'. (Exit code: 1)
make[1]: *** [rts/ghc.mk:345: rts/dist/build/StgCRun.o] Error 1
make: *** [Makefile:128: all] Error 2
```

**Comment by paparodeo** [comment]
`.cfi_startproc` needs to come after the label not before. we have a number of fixes for this: eg https://github.com/NixOS/nixpkgs/pull/354734/files tho IIRC fix for ghc is not as simple.

**Comment by sternenseemann** [comment]
I figured, too, but I have not been able to figure out why GHC >= 9.6 succeeds. It does not look like they've actually reordered anything upstream from diffing the obvious `rts` source files between 9.4.8 to 9.6.6.

**Comment by sternenseemann** [comment]
GHC ticket: <https://gitlab.haskell.org/ghc/ghc/-/issues/25608>

**Comment by domenkozar** [comment]
@sternenseemann couldn't we downgrade llvm for ghc? This is blocking us on a few projects

**Comment by sternenseemann** [comment]
LLVM is not the problem and would be trivial to downgrade, the problem (as far as I can tell) concerns clang which is used as assembler on darwin. We inherit it from `stdenv` in various ways which makes this very annoying and ugly to implement. It would be acceptable in principle, though.

**Comment by dtomvan** [comment]
Can't `stdenv` get overridden in https://github.com/NixOS/nixpkgs/blob/606918626f7bd3ba60be60ee817c062d3c3eb3b9/pkgs/top-level/haskell-packages.nix?plain=1#L151? Like
`stdenv = if stdenv.buildPlatform.isDarwin then pkgs.llvmPackages_12.stdenv else stdenv;`

**Comment by sternenseemann** [comment]
No, this is not enough which my message was about. If you look at the actual package expression you'll see that this override would not affect all uses of stdenv.

**Comment by PedroHLC** [comment]
Anyone has a workaround/overlay?

**Comment by sternenseemann** [comment]
I was able to diagnose the issue. Fixing it on the Haskell side is easy for GHC 9.4 which fixes that version of GHC on x86_64-darwin (only). Unfortunately, the relevant patch is nontrivial to backport to GHC  < 9.2 which hasn't been done yet. 

#410085.

**Comment by miniBill** [comment]
When https://github.com/avh4/elm-format/pull/829 lands then `elm-format` will be fixed

**Comment by sternenseemann** [comment]
I've opened a [clang bug report for this](https://github.com/llvm/llvm-project/issues/148170), in case LLVM would consider this a bug. Even if they do, it'll probably not help us much since it'd be a while until we upgrade to LLVM 21/22.

My feeling is that the best way forward is to try and port everything relevant to GHC >= 9.4 and to drop GHC < 9.4 which our [new policy would allow](https://github.com/NixOS/nixpkgs/pull/409512).

If someone wants to backport the fixes to < 9.4, be my guest, but it looks to be very hard, unfortunately.

**Comment by sternenseemann** [comment]
We no longer build any GHC < 9.4 from source.
